pkg/quarterdeck/invites.go

```diff
@@ -198,3 +202,140 @@ func (s *Server) InviteCreate(c *gin.Context) {
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve user from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	// User must be verified to accept the invite
+	if !user.EmailVerified {
+		log.Debug().Msg("user has not verified their email address")
+		c.JSON(http.StatusForbidden, api.ErrorResponse(responses.ErrVerifyEmail))
+		return
+	}
+
+	// Accept the invite and handle errors
+	if s.acceptInvite(c, user, c.Param("token")) != nil {
+		return
+	}
+
+	// Construct the claims for the user
+	var newClaims *tokens.Claims
+	if newClaims, err = user.NewClaims(ctx); err != nil {
+		sentry.Error(c).Err(err).Msg("could not create new claims for user")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	out := &api.LoginReply{
+		LastLogin: user.LastLogin.String,
+	}
+	if out.AccessToken, out.RefreshToken, err = s.tokens.CreateTokenPair(newClaims); err != nil {
+		sentry.Error(c).Err(err).Msg("could not create access and refresh token")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	// Update the user's last login in a Go routine
+	s.tasks.QueueContext(sentry.CloneContext(c), tasks.TaskFunc(func(ctx context.Context) error {
+		ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+		defer cancel()
+		return user.UpdateLastLogin(ctx)
+	}), tasks.WithError(fmt.Errorf("could not update last login timestamp for user %s", user.ID.String())))
+
+	c.JSON(http.StatusOK, out)
+}
+
+// acceptInvite is a helper method that accepts an invitation and updates the user's
+// organization membership in the database. This method handles the logging and error
+// responses if the invitation is invalid or there is a database error.
+func (s *Server) acceptInvite(c *gin.Context, user *models.User, token string) (err error) {
+	// Retrieve the invitation from the database
+	var invite *models.UserInvitation
+	if invite, err = models.GetUserInvite(c.Request.Context(), token); err != nil {
+		if errors.Is(err, models.ErrNotFound) {
+			log.Debug().Msg("could not find invite token")
+			c.JSON(http.StatusBadRequest, api.ErrorResponse(responses.ErrRequestNewInvite))
+			metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "invite token not found").Inc()
+			return err
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve the user invite from the database")
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "could not get invite token").Inc()
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return err
+	}
+
+	// Verify that the invite is for this email address
+	if err = invite.Validate(user.Email); err != nil {
+		switch {
+		case errors.Is(err, db.ErrTokenExpired):
+			log.Debug().Msg("invite token expired")
+			metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "invite token expired").Inc()
```

Comment:
Not sure if this is going to break the metrics, these strings are a bit different than they were in Login.